stdlib/3/os/path.pyi

```diff
@@ -75,29 +130,41 @@ if sys.version_info < (3, 0):
     def join(__p1: bytes, __p2: bytes, __p3: Text, *p: _PathType) -> Text: ...
     @overload
     def join(__p1: bytes, __p2: bytes, __p3: bytes, __p4: Text, *p: _PathType) -> Text: ...
+elif sys.version_info >= (3, 6):
+    # Mypy complains that the signatures overlap (same for relpath below), but things seem to behave correctly anyway.
+    @overload
+    def join(path: _StrPath, *paths: _StrPath) -> Text: ...  # type: ignore
+    @overload
+    def join(path: _BytesPath, *paths: _BytesPath) -> bytes: ...
 else:
     def join(path: AnyStr, *paths: AnyStr) -> AnyStr: ...
 
-def normcase(path: AnyStr) -> AnyStr: ...
-def normpath(path: AnyStr) -> AnyStr: ...
-if sys.platform == 'win32':
-    def realpath(path: AnyStr) -> AnyStr: ...
-else:
-    def realpath(filename: AnyStr) -> AnyStr: ...
-def relpath(path: AnyStr, start: _PathType = ...) -> AnyStr: ...
+@overload
+def relpath(path: _StrPath, start: Optional[_StrPath] = ...) -> Text: ...  # type: ignore
+@overload
+def relpath(path: _BytesPath, start: _BytesPath) -> bytes: ...
```

Comment:
`relpath` isn't guarded by a 3.6 version check and there is no AnyStr fallback.

Also, the bytes variant is missing `Optional`.

Reply:
That is intentional. `_StrPath` is just `Text` below 3.6, so this implementation is correct there.

If start is None, the code uses os.curdir, which is Text, so the path also has to be Text. Although come to think of it, we may need something more complicated in Python 2, where you can mix bytes and Text paths.